### PR TITLE
feat: add searcher-specific InvalidHP logic [DET-4334, DET-4335]

### PR DESCRIPTION
**Pull request by naren-determined**
## Description

Add searcher logic to handle InvalidHP exceptions, and implement searcher-specific handling for ASHA and other searchers.

## Test Plan

Run an ASHA experiment and randomly pause/restart/terminate trials and ensure that progress does not break (i.e. go over 100%) and the experiment behaves as normal.

## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

File: master/pkg/searcher/event_log.go [hunk 1]
```diff
@@ -70,13 +72,18 @@ func (el *EventLog) TrialCreated(create Create, trialID int) {
 	el.uncommitted = append(el.uncommitted, trialCreated)
 	el.TrialIDs[create.RequestID] = trialID
 	el.RequestIDs[trialID] = create.RequestID
+	el.unitsCompletedByTrial[create.RequestID] = 0
```
Comment (stoksc) [inline master/pkg/searcher/event_log.go:75]: blocking: maybe`delete(el.unitsCompletedByTrial, create.RequestID)` in `TrialClosed` to save ourselves a bit?

File: master/pkg/searcher/asha.go [hunk 2]
```diff
@@ -239,12 +239,36 @@ func (s *asyncHalvingSearch) progress(float64) float64 {
 	if allTrials == s.maxTrials {
 		progress = math.Max(float64(s.trialsCompleted)/float64(s.maxTrials), progress)
 	}
+	// Cap progress at 95% for InvalidHP cases
+	if progress > 1 {
+		progress = float64(0.95)
```
Comment (stoksc) [inline master/pkg/searcher/asha.go:244]: blocking: since we're correcting the progress, to we need this?

File: master/pkg/searcher/event_log.go [hunk 3]
```diff
@@ -70,13 +72,18 @@ func (el *EventLog) TrialCreated(create Create, trialID int) {
...
 }
 
 // TrialExitedEarly marks the trial with the given requestID as exited early.
-func (el *EventLog) TrialExitedEarly(requestID RequestID) {
+func (el *EventLog) TrialExitedEarly(requestID RequestID, exitedReason workload.ExitedReason) {
 	// If we are exiting early and we haven't seen this exiting early
 	// message before, return true to send the message down to the search
 	// method.
+	if exitedReason == workload.InvalidHP {
+		unitsCancelled := el.unitsCompletedByTrial[requestID]
```
Comment (stoksc) [inline master/pkg/searcher/event_log.go:84]: blocking: might as well delete this from the map here too.

File: master/pkg/searcher/asha.go [hunk 4]
```diff
@@ -239,12 +239,36 @@ func (s *asyncHalvingSearch) progress(float64) float64 {
...
+	if exitedReason == workload.InvalidHP {
+		var ops []Operation
+		s.earlyExitTrials[requestID] = true
+		ops = append(ops, NewClose(requestID))
+		s.closedTrials[requestID] = true
+		// Remove metrics associated with InvalidHP trial across all rungs
+		highestRungIndex := s.trialRungs[requestID]
+		for rungIndex := 0; rungIndex <= highestRungIndex; rungIndex++ {
+			rung := s.rungs[rungIndex]
+			rung.metrics = nil
```
Comment (stoksc) [inline master/pkg/searcher/asha.go:261]: blocking: here don't we need to iterate over `rung.metrics` and just delete the metric where `metric.requestID == requestID`, not just nil out the metrics for the whole rung, right?

File: master/pkg/searcher/searcher.go [hunk 5]
```diff
@@ -66,7 +66,7 @@ func (s *Searcher) TrialExitedEarly(
 		return nil, errors.Errorf("unexpected trial ID sent to searcher: %d", trialID)
 	}
 
-	s.eventLog.TrialExitedEarly(requestID)
+	s.eventLog.TrialExitedEarly(requestID, *exitedReason)
```
Comment (stoksc) [inline master/pkg/searcher/searcher.go:69]: blocking: I know this only gets called when the trial actor sees the `exitedReason` is non nil, so this being nil is a would only error be a programming error, but can we check it and `panic` with a really specific message if that happens?

File: master/pkg/searcher/asha.go (hunk 2 above)
Comment (naren-determined) [reply master/pkg/searcher/asha.go:244]: So in the case of ASHA specifically, correcting progress would be difficult as we can’t know whether a trial is closed forever until we finish all possible promotions in a rung (which means trials can start and stop somewhat arbitrarily), estimating progress by units completed is not as feasible. As such, I'm hard capping progress here: see https://determined-ai.slack.com/archives/C019YGJQ0P3/p1607542853041000 for relevant slack discussion.

File: master/pkg/searcher/event_log.go (hunk 3 above)
Comment (stoksc) [reply master/pkg/searcher/event_log.go:84]: yep, i forgot this, even after a debugged a bug related to it lol

File: master/pkg/searcher/asha.go (hunk 2 above)
Comment (stoksc) [reply master/pkg/searcher/asha.go:244]: yeah, forgot asha progress is totally different for a minute

Comment (stoksc) [reply master/pkg/searcher/asha.go:244]: well that being said, can you drop a descriptive comment explaining why?

Comment (naren-determined) [reply master/pkg/searcher/asha.go:244]: Yes good idea

File: master/pkg/searcher/searcher.go [hunk 6]
```diff
@@ -65,7 +66,9 @@ func (s *Searcher) TrialExitedEarly(
 	if !ok {
 		return nil, errors.Errorf("unexpected trial ID sent to searcher: %d", trialID)
 	}
-
+	if exitedReason == nil {
+		panic(fmt.Sprintf("Trial %d exited early with ExitedReason nil.\n", trialID))
```
Comment (stoksc) [inline master/pkg/searcher/searcher.go:70]:
```suggestion
		panic(fmt.Sprintf("trial %d exited early with ExitedReason nil", trialID))
```

File: master/pkg/searcher/asha.go [hunk 7]
```diff
@@ -258,7 +259,11 @@ func (s *asyncHalvingSearch) trialExitedEarly(
 		highestRungIndex := s.trialRungs[requestID]
 		for rungIndex := 0; rungIndex <= highestRungIndex; rungIndex++ {
 			rung := s.rungs[rungIndex]
-			rung.metrics = nil
+			for i, trialMetric := range rung.metrics {
+				if trialMetric.requestID == requestID {
+					rung.metrics = append(rung.metrics[:i], rung.metrics[i+1:]...)
```
Comment (stoksc) [inline master/pkg/searcher/asha.go:264]: I think we can `break` after this right, there should only be one metric per trial per rung.

**Review by stoksc — COMMENTED** [review]
do you think we can add a test to cover asha since it's relatively complicated.

**Comment by stoksc** [comment]
@naren-determined any thoughts on a test?

File: master/pkg/searcher/asha.go [hunk 8]
```diff
@@ -237,14 +238,42 @@ func (s *asyncHalvingSearch) progress(float64) float64 {
 	// Give ourselves an overhead of 20% of maxTrials when calculating progress.
 	progress := float64(allTrials) / (1.2 * float64(s.maxTrials))
 	if allTrials == s.maxTrials {
```
Comment (liamcli) [inline master/pkg/searcher/asha.go:240]: we should account for invalidTrials in line 199 as well.

**Review by liamcli — APPROVED** [review]

**Review by stoksc — APPROVED** [review]
lgtm

**Pull request merged**